File: Assignments/01._Introduction_Data_Formats/01a [Individual] Data parsing servers - Part I/scripts/parse.py
import json
import yaml
import csv
import xml.etree.ElementTree as ET
import os

# Set the base directory to go up one level to find the "data" folder
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))


# Read and parse a text file
def parse_txt(file_path):
    with open(file_path, "r") as f:
        data = {}
        for line in f:
            key, value = line.strip().split(": ")
            data[key] = value
    return data

# Read and parse a JSON file
def parse_json(file_path):
    with open(file_path, "r") as f:
        return json.load(f)

# Read and parse an XML file
def parse_xml(file_path):
    tree = ET.parse(file_path)
    root = tree.getroot()
    data = []
    for user in root.findall("user"):
        data.append({
            "name": user.find("name").text,
            "age": int(user.find("age").text),
            "location": user.find("location").text
        })
    return data

# Read and parse a YAML file
def parse_yaml(file_path):
    with open(file_path, "r") as f:
        return yaml.safe_load(f)

# Read and parse a CSV file
def parse_csv(file_path):
    with open(file_path, newline="") as f:
        reader = csv.DictReader(f)
        return [row for row in reader]

# File paths
files = {
    "txt": os.path.join(BASE_DIR, "data", "data.txt"),
    "json": os.path.join(BASE_DIR, "data", "data.json"),
    "xml": os.path.join(BASE_DIR, "data", "data.xml"),
    "yaml": os.path.join(BASE_DIR, "data", "data.yaml"),
    "csv": os.path.join(BASE_DIR, "data", "data.csv"),
}

# Parsing all files
print("TXT:", parse_txt(files["txt"]))
print("JSON:", parse_json(files["json"]))
print("XML:", parse_xml(files["xml"]))
print("YAML:", parse_yaml(files["yaml"]))
print("CSV:", parse_csv(files["csv"]))

File: Assignments/01._Introduction_Data_Formats/01a [Individual] Data parsing servers - Part I/scripts/parse.js
const fs = require("fs");
const yaml = require("js-yaml");
const xml2js = require("xml2js");
const path = require("path"); // Import path module

// Get the absolute path to the "data" folder
const BASE_DIR = path.join(__dirname, "..", "data");

const files = {
    txt: path.join(BASE_DIR, "data.txt"),
    json: path.join(BASE_DIR, "data.json"),
    xml: path.join(BASE_DIR, "data.xml"),
    yaml: path.join(BASE_DIR, "data.yaml"),
    csv: path.join(BASE_DIR, "data.csv")
};


// Read and parse a text file
function parseTxt(filePath) {
    const data = fs.readFileSync(filePath, "utf-8").split("\n");
    return Object.fromEntries(data.map(line => line.split(": ").map(s => s.trim())));
}

// Read and parse a JSON file
function parseJson(filePath) {
    return JSON.parse(fs.readFileSync(filePath, "utf-8"));
}

// Read and parse an XML file
async function parseXml(filePath) {
    const data = fs.readFileSync(filePath, "utf-8");
    const parsedData = await xml2js.parseStringPromise(data, { explicitArray: false }); // Prevent unnecessary arrays
    return parsedData.users.user.map(user => ({
        name: user.name,
        age: parseInt(user.age, 10),  // Convert age to an integer
        location: user.location
    }));
}


// Read and parse a YAML file
function parseYaml(filePath) {
    return yaml.load(fs.readFileSync(filePath, "utf-8"));
}

// Read and parse a CSV file
function parseCsv(filePath) {
    const data = fs.readFileSync(filePath, "utf-8").split("\n");
    const headers = data[0].split(",");
    return data.slice(1).map(row => Object.fromEntries(row.split(",").map((val, i) => [headers[i], val.trim()])));
}

// Run parsing functions
console.log("TXT:", parseTxt(files.txt));
console.log("JSON:", parseJson(files.json));

parseXml(files.xml).then(xmlData => {
    console.log("XML:", xmlData);
});

console.log("YAML:", parseYaml(files.yaml));
console.log("CSV:", parseCsv(files.csv));
